common/ipmi/oem_1s_handler.c

```diff
@@ -0,0 +1,864 @@
+		// notice BMC about sensor temporary in not accessible status
+		msg->data[2] = (snr_report_status | 0x20);
+		msg->data_len = 3;
+		msg->completion_code = CC_SUCCESS;
+		break;
+	case SNR_POLLING_DISABLE:
+		// getting sensor cache while sensor polling disable
+		msg->completion_code = CC_SENSOR_NOT_PRESENT;
+		break;
+	case SNR_FAIL_TO_ACCESS:
+		// transection error
+		msg->completion_code = CC_NODE_BUSY;
+		break;
+	case SNR_NOT_FOUND:
+		// request sensor number not found
+		msg->completion_code = CC_INVALID_DATA_FIELD;
+		break;
+	case SNR_UNSPECIFIED_ERROR:
+	default:
+		// unknown error
+		msg->completion_code = CC_UNSPECIFIED_ERROR;
+		break;
+	}
+	return;
+}
+
+__weak void OEM_1S_GET_SET_GPIO(ipmi_msg *msg)
+{
+	uint8_t completion_code = CC_INVALID_LENGTH;
+	uint8_t gpio_num = gpio_ind_to_num_table[msg->data[1]];
+
+	if (msg->data[0] == 0) { // Get GPIO output status
+		if (msg->data_len == 2) {
+			msg->data[0] = gpio_num;
+			msg->data[1] = gpio_get(gpio_num);
+			completion_code = CC_SUCCESS;
+		}
+	} else if (msg->data[0] == 1) { // Set GPIO output status
+		if (msg->data_len == 3) {
+			msg->data[0] = gpio_num;
+			gpio_conf(gpio_num, GPIO_OUTPUT);
+			gpio_set(gpio_num, msg->data[2]);
+			msg->data[1] = gpio_get(gpio_num);
+			completion_code = CC_SUCCESS;
+		}
+	} else if (msg->data[0] == 2) { // Get GPIO direction status
+		completion_code = CC_NOT_SUPP_IN_CURR_STATE;
+	} else if (msg->data[0] == 3) { // Set GPIO direction status
+		if (msg->data_len == 3) {
+			if (msg->data[2]) {
+				gpio_conf(gpio_num, GPIO_OUTPUT);
+			} else {
+				gpio_conf(gpio_num, GPIO_INPUT);
+			}
+			msg->data[0] = gpio_num;
+			msg->data[1] = msg->data[2];
+			completion_code = CC_SUCCESS;
+		}
+	}
+
+	if (completion_code != CC_SUCCESS) {
+		msg->data_len = 0;
+	} else {
+		msg->data_len = 2; // Return GPIO number, status
+	}
+	msg->completion_code = completion_code;
+	return;
+}
+
+__weak void OEM_1S_I2C_DEV_SCAN(ipmi_msg *msg)
+{
+	if ((msg->data[0] == 0x9C) && (msg->data[1] == 0x9C) && (msg->data[2] == 0x00)) {
+		while (1)
+			; // hold firmware for debug only
+	}
+
+	if (msg->data_len != 1) { // only input scan bus
+		msg->completion_code = CC_INVALID_LENGTH;
+		return;
+	}
+
+	uint8_t bus = i2c_bus_to_index[msg->data[0]];
+
+	i2c_scan(bus, &msg->data[0], (uint8_t *)&msg->data_len);
+
+	msg->completion_code = CC_SUCCESS;
+	return;
+}
+
+__weak void OEM_1S_GET_BIC_STATUS(ipmi_msg *msg)
+{
+	if (msg->data_len != 0) {
+		msg->completion_code = CC_INVALID_LENGTH;
+		return;
+	}
+
+	msg->data[0] = FIRMWARE_REVISION_1;
+	msg->data[1] = FIRMWARE_REVISION_2;
+
+	msg->data_len = 2;
+	msg->completion_code = CC_SUCCESS;
+	return;
+}
+
+__weak void OEM_1S_RESET_BIC(ipmi_msg *msg)
+{
+	if (msg->data_len != 0) {
+		msg->completion_code = CC_INVALID_LENGTH;
+		return;
+	}
+
+	submit_bic_warm_reset();
+
+	msg->data_len = 0;
+	msg->completion_code = CC_SUCCESS;
+	return;
+}
+
+__weak void OEM_1S_12V_CYCLE_SLOT(ipmi_msg *msg)
+{
+	if (msg->data_len != 0) {
```

Comment:
msg == NULL check?

Reply:
We have fixed it and done the forced update.
Please review it. Thanks.